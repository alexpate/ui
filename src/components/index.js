/**
 * Typography
 */
export { H1 } from './typography/h1';
export { H2 } from './typography/h2';
export { H3 } from './typography/h3';
export { H4 } from './typography/h4';
export { Text } from './typography/text';
export { Heading } from './typography/heading';

export { Button } from './button';
export { ButtonGroup } from './button-group';
export { FieldHint } from './field-hint';
export { Icon } from './icon';
export { Input } from './input';
export { List } from './list';
export { LoadingSpinner } from './loading-spinner';
export { Radio } from './radio';
export { StarRating } from './star-rating';
export { StepMarks } from './step-marks';
export { ToggleButton } from './toggle-button';
