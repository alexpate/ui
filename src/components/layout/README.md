# Layout

The layout components are the primitive building blocks of the Kalo User Interface. They are a combination of `grid-styled`, and some custom rolled components.

## Box and Flex (via `grid-styled`)
A wrapper component that implements some default flex values.

## Inline
An `inline-block` element.

## Spacing
Margin and padding can be set via the `margin` and `padding` props. There are also more granular props for specifycing particular directions: `marginLeft`, `paddingTop` etc.

These props can take a variety of values:
- Value specific values: `padding="20px"`
- Or scale values: `margin={2}`

The scale values relate to the 8px grid. Any value between 1 and 8, correlates to a mutiple of 8px:

```js
marginLeft={1} // margin-left: 8px
marginLeft={2} // margin-left: 16px
marginLeft={3} // margin-left: 24px
```

## The css and style props
The `css` prop is inherited from Emotion and is available on every component. Styles passed in through this prop, will be set via a classname generated at run-time.

The `style` prop is used to add styles to an element inline via the style attribute. This should be used for regulary computed styles as it is considerably more performant.