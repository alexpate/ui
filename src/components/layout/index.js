export {
  Box,
  Flex,
  Block,
  Inline,
  InlineBlock,
  InlineFlex,
  A
} from './layout';
