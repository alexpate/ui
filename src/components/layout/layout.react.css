.ui-box {
  position: relative;
  flex-direction: column;
  align-items: stretch;
  flex-shrink: 0;
  align-content: flex-start;
  display: flex;
}

.ui-flex {
  display: flex;
}

.ui-block {
  display: block;
}

.ui-inline {
  display: inline-block;
  vertical-align: bottom;
}

.ui-inline-flex {
  display: inline-flex;
}

.ui-a {
  display: inline-flex;
}
