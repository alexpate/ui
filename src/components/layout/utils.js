import React, {PureComponent} from 'react';
import {parseStyleProps, cleanProps} from '../../utils/style';
import cx from 'classnames';

export const makePrimitive = (name, DefaultComponent, defaultStyle) => {
  class Primitive extends PureComponent {
    render() {
      const {
        component: Component = DefaultComponent,
        elRef,
        style: propStyle,
        className,
        ...otherProps
      } = this.props;

      const {props, style} = parseStyleProps({
        ...otherProps,
      });

      const cleanedProps = cleanProps(props);

      return (
        <Component
          ref={elRef}
          className={className ? cx(className) : null}
          css={{...defaultStyle, ...style}}
          {...cleanedProps}
          style={{...propStyle}}
        />
      );
    }
  }

  Primitive.displayName = `Primitive${name}`;
  return Primitive;
};
