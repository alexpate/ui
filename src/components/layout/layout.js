/* @flow */
import {spaceProps} from 'utils/style';
import styled from 'react-emotion';

/**
 * Layout primitives
 */

export const Box = styled.div`
  flex-direction: column;
  align-items: stretch;
  flex-shrink: 0;
  align-content: flex-start;
  display: flex;
  ${spaceProps};
`;
Box.displayName = 'Box';

export const Flex = styled.div`
  display: flex;
  flex-direction: row;
  display: flex;
  ${spaceProps};
`;
Flex.displayName = 'Flex';

export const Block = styled.div`
  display: block;
  ${spaceProps};
`;
Block.displayName = 'Block';

export const Inline = styled.span`
  display: inline-block;
  vertical-align: bottom;
  ${spaceProps};
`;
Inline.displayName = 'Inline';

export const InlineFlex = styled.span`
  display: inline-flex;
  ${spaceProps};
`;
InlineFlex.displayName = 'InlineFlex';

/**
 * All of our html primitives should follow here so our api is consistent to the very root
 * Eg. Button, Input, Select, Table etc.
 */
export const A = styled('a')`display: inline-block;`;
A.displayName = 'A';

export const H1 = styled.h1`display: block;`;
H1.displayName = 'H1';

export const H2 = styled.h2`display: block;`;
H2.displayName = 'H2';

export const H3 = styled.h3`display: block;`;
H3.displayName = 'H3';

export const H4 = styled.h4`display: block;`;
H4.displayName = 'H4';
