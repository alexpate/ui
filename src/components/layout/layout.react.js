import {makePrimitive} from './utils';
import styles from './layout.react.css';
/**
 * Layout primitives
 */

export const Box = makePrimitive('Box', 'div', styles['ui-box']);
Box.displayName = 'Box';

export const Flex = makePrimitive('Flex', 'div', styles['ui-flex']);
Flex.displayName = 'Flex';

export const Block = makePrimitive('Block', 'div', styles['ui-block']);
Block.displayName = 'Block';

export const Inline = makePrimitive('Inline', 'span', styles['ui-inline']);
Inline.displayName = 'Inline';

export const InlineFlex = makePrimitive(
  'InlineFlex',
  'span',
  styles['ui-inline-flex']
);
InlineFlex.displayName = 'InlineFlex';

/**
 * All of our html primitives should follow here so our api is consistent to the very root
 * Eg. Button, Input, Select, Table etc.
 */

export const A = makePrimitive('A', 'a', styles['ui-a']);
A.displayName = 'A';

/**
 * This is an empty component that provides backwards compatibility as we phase out the
 * layout primivites above. Its sole purpose is is to convert inline styles props in to
 * actual CSS.
 */

export const UIBase = makePrimitive('UIBase', 'div');
UIBase.displayName = 'UIBase';
