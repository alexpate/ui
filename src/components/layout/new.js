/* @flow */
import React from 'react';
import {parseStyleAndSpacingProps} from 'utils/style';
import {cx} from 'emotion';

import {Box as GridBox, Flex as GridFlex} from 'grid-emotion';

function extendPrimitive(name, DefaultComponent, display, defaultStyle) {
  return (initialProps: any) => {
    const {
      component: Component = DefaultComponent,
      elRef,
      style: propStyle,
      className,
      ...otherProps
    } = initialProps;

    const {props, spacingProps, style} = parseStyleAndSpacingProps({
      ...defaultStyle,
      ...otherProps,
    });

    return (
      <Component
        ref={elRef}
        className={cx(className)}
        css={{...style, ...propStyle}}
        {...spacingProps}
        {...props}
      />
    );
  };
}

export const Box = extendPrimitive('Box', GridBox, 'flex', {
  position: 'relative',
  flexDirection: 'column',
  alignItems: 'stretch',
  flexShrink: 0,
  alignContent: 'flex-start',
  display: 'flex',
});

export const Flex = extendPrimitive('Flex', GridFlex, 'flex', {
  flexDirection: 'row',
  display: 'flex',
});

export const Block = extendPrimitive('Block', 'div', 'block');
export const Inline = extendPrimitive('Inline', 'span', 'inline-block', {
  verticalAlign: 'bottom',
});
