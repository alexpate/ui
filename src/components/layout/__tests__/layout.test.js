/* @flow */
/* eslint-env jest */
import {shallow} from 'enzyme';
import {testComponent} from 'utils/test/react';
import serializer from 'enzyme-to-json/serializer';

import {Box, Block, Flex, Inline, InlineFlex} from 'components/layout';

expect.addSnapshotSerializer(serializer);

describe('components/layout', () => {
  describe('Box', () => {
    const create = testComponent(Box, (props = {}) => props);

    it('should render correctly', () => {
      const {element} = create({
        flex: 0,
        onClick: null,
      });
      const result = shallow(element);
      expect(result).toMatchSnapshot();
    });

    it('should allow overriding of default styles', () => {
      const {element} = create({
        flex: 0,
        alignContent: 'flex-end',
        flexShrink: 1,
        style: {color: 'red'},
      });
      const result = shallow(element);
      expect(result).toMatchSnapshot();
    });

    it('should allow passing vector style values', () => {
      const {element} = create({
        margin: [0, 5, '10%', 15],
        padding: [15, undefined],
      });
      const result = shallow(element);
      expect(result).toMatchSnapshot();
    });
  });

  describe('Block', () => {
    const create = testComponent(Block, (props = {}) => props);

    it('should render correctly', () => {
      const {element} = create({
        flex: 0,
        onClick: null,
      });
      const result = shallow(element);
      expect(result).toMatchSnapshot();
    });
  });

  describe('Flex', () => {
    const create = testComponent(Flex, (props = {}) => props);

    it('should render correctly', () => {
      const {element} = create({
        flex: 0,
        onClick: null,
      });
      const result = shallow(element);
      expect(result).toMatchSnapshot();
    });
  });

  describe('Inline', () => {
    const create = testComponent(Inline, (props = {}) => props);

    it('should render correctly', () => {
      const {element} = create({
        flex: 0,
        onClick: null,
      });
      const result = shallow(element);
      expect(result).toMatchSnapshot();
    });
  });

  describe('InlineFlex', () => {
    const create = testComponent(InlineFlex, (props = {}) => props);

    it('should render correctly', () => {
      const {element} = create({
        flex: 0,
        onClick: null,
      });
      const result = shallow(element);
      expect(result).toMatchSnapshot();
    });
  });
});
