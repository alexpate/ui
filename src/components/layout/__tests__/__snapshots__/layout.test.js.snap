// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Block should render correctly 1`] = `
<div
  className="ui-block"
  onClick={null}
  style={
    Object {
      "flex": 0,
    }
  }
/>
`;

exports[`Box should allow overriding of default styles 1`] = `
<div
  className="ui-box"
  style={
    Object {
      "alignContent": "flex-end",
      "color": "red",
      "flex": 0,
      "flexShrink": 1,
    }
  }
/>
`;

exports[`Box should allow passing spacing scales 1`] = `
<div
  className="ui-box"
  style={
    Object {
      "margin": "8px 16px 24px",
      "paddingTop": "8px",
    }
  }
/>
`;

exports[`Box should allow passing vector style values 1`] = `
<div
  className="ui-box"
  style={
    Object {
      "margin": "0px 5px 10% 15px",
      "padding": "15px ",
    }
  }
/>
`;

exports[`Box should render correctly 1`] = `
<div
  className="ui-box"
  style={Object {}}
/>
`;

exports[`Flex should render correctly 1`] = `
<div
  className="ui-flex"
  onClick={null}
  style={
    Object {
      "flex": 0,
    }
  }
/>
`;

exports[`Inline should render correctly 1`] = `
<span
  className="ui-inline"
  onClick={null}
  style={
    Object {
      "flex": 0,
    }
  }
/>
`;

exports[`InlineFlex should render correctly 1`] = `
<span
  className="ui-inline-flex"
  onClick={null}
  style={
    Object {
      "flex": 0,
    }
  }
/>
`;
